File: C++/splitArray.cpp
/**
 * @File    : splitArray.cpp
 * @Brief   : 分割数组的最大值
 * @Link    : https://leetcode-cn.com/problems/split-array-largest-sum/
 * @Author  : Wei Li
 * @Date    : 2021-08-25
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    bool check(std::vector<int> &nums, int x, int m)
    {
        long long sum = 0;
        int cnt = 1;
        for (int i = 0; i < nums.size(); i++)
        {
            if (sum + nums[i] > x)
            {
                cnt++;
                sum = nums[i];
            }
            else
            {
                sum += nums[i];
            }
        }
        return cnt <= m;
    }

    int splitArray(std::vector<int> &nums, int m)
    {
        long long left = 0, right = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            right += nums[i];
            if (left < nums[i])
            {
                left = nums[i];
            }
        }
        while (left < right)
        {
            long long mid = (left + right) >> 1;
            if (check(nums, mid, m))
            {
                right = mid;
            }
            else
            {
                left = mid + 1;
            }
        }
        return left;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{1, 2, 3, 4, 5};
    // std::vector<int> nums{7, 2, 5, 10, 8};
    int m = 2;

    Solution solution;
    auto split_array_result = solution.splitArray(nums, m);
    std::cout << "The solution of this problem is " << split_array_result << std::endl;

    return 0;
}

File: Python/splitArray.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: splitArray.py
@Function: 分割数组的最大值
@Link: https://leetcode-cn.com/problems/split-array-largest-sum/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-25
"""


from typing import List


class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        def check(x: int) -> bool:
            total, cnt = 0, 1
            for num in nums:
                if total + num > x:
                    cnt += 1
                    total = num
                else:
                    total += num
            return cnt <= m

        left = max(nums)
        right = sum(nums)
        while left < right:
            mid = (left + right) // 2
            if check(mid):
                right = mid
            else:
                left = mid + 1

        return left


# -------------------------
if __name__ == "__main__":
    nums = [1,4,4]
    m = 3

    solution = Solution()
    split_array_result = solution.splitArray(nums, m)
    print(f"The solution of this problem is {split_array_result}")
